File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.java
class Solution {
    public int search(int[] nums, int target) {
        int r = nums.length - 1;
        int l = 0;
        while (l <= r) {
            int mid = (r + l) / 2;
            if (nums[mid] == target) {
                return mid;
            }

            if (nums[l] <= nums[mid]) {
                if (nums[l] > target || nums[mid] < target) {
                l = mid + 1;
                } else {
                    r = mid - 1;
                }
            } else {
                if (target < nums[mid] || target > nums[r]) {
                    r = mid - 1;
                } else {
                    l = mid + 1;
                }
            }
            
        }
        return -1;
    }
}

File: Data Structures & Algorithms/search-2d-matrix/submission-2.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        left = 0
        right = len(matrix) - 1

        while left <= right:
            mid = (left + right) // 2
            if target < matrix[mid][0]:
                right = mid - 1
            elif target > matrix[mid][-1]:
                left = mid + 1
            else:
                break

        if not (left <= right):
            return False

        index = (left+right) //2
        array = matrix[index]
        inner_r = len(array) - 1
        inner_l = 0
        while inner_l <= inner_r:
            inner_m = (inner_r + inner_l) // 2
            if array[inner_m] == target:
                return True
            elif array[inner_m] > target:
                inner_r = inner_m -1 
            else:
                inner_l = inner_m + 1

        return False

